image-customize

```diff
@@ -82,6 +83,114 @@ class InstallAction(ActionBase):
         machine_instance.execute(f"{install_command} {package}", timeout=1800)
 
 
+class BuildAction(ActionBase):
+    '''Build and install distribution package(s) from dist tarball'''
+
+    quick = False
+    verbose = False
+
+    @staticmethod
+    def execute(machine_instance, tarball):
+        # upload the tarball
+        tarname = os.path.basename(tarball)
+        vm_tar = os.path.join("/var/tmp", tarname)
+        # looks like cockpit-foo-123.10.gdbde381.tar.xz
+        m = re.search(r'([^-]+)\.tar(\..*)?$', tarname)
+        if not m:
+            raise ValueError(f"cannot parse version number from tarball {tarname}")
+        version = m.group(1)
+        machine_instance.upload([tarball], vm_tar, relative_dir=".")
+
+        machine_instance.execute("rm -rf /var/tmp/build; mkdir -p /var/tmp/build")
+
+        # this will fail if neither is available -- exception is clear enough, this is a developer tool
+        out = machine_instance.execute("(which pbuilder || which mock || which pacman) 2>/dev/null")
+        if 'pbuilder' in out:
+            BuildAction.build_deb(machine_instance, vm_tar, version)
+        elif 'mock' in out:
+            BuildAction.build_rpm(machine_instance, vm_tar, version)
+        elif 'pacman' in out:
+            BuildAction.build_arch(machine_instance, vm_tar, version)
+        else:
+            raise NotImplementedError(f"unknown build platform: {out}")
+
+        # machine_instance.execute(f"{install_command} {package}", timeout=1800)
+
+    def build_deb(machine, vm_tar, version):
+        build_opts = 'nocheck' if BuildAction.quick else ''
+
+        # build source packge
+        machine.execute(f"""
+            set -eu
+            export DEB_BUILD_OPTIONS='{build_opts}'
+            tar -C /var/tmp/build -xf '{vm_tar}'
+            cd "$(ls -d /var/tmp/build/*)"
+            # find and copy debian packaging directory
+            cp -r "$(dirname $(find -path '*/debian/control'))" .
+            # create orig.tar link for building dsc
+            source=$(awk '/^Source: / {{ print $2 }}' debian/control)
+            ln -s '{vm_tar}' ../${{source}}_{version}.orig.tar.xz
+            # fill in the version number
+            sed -i '1 s/([^)-]*/({version}/' debian/changelog
```

Comment:
I wonder if this is another case of something we could do statically in the tarball itself, to prevent the need for after-the-fact patching, just as we recently did for the spec?  Definitely for another PR, though.

Reply:
That would make rpm vs. spec more consistent; there was some reason why I didn't do that back then, but I can't remember. I added it to my TODO list to try and put that into the dist tarball.